lib/auth/grpcserver.go

```diff
@@ -1603,26 +1618,115 @@ func (g *GRPCServer) GenerateUserSingleUseCerts(stream proto.AuthService_Generat
 	return nil
 }
 
-func validateUserSingleUseCertRequest(req *proto.UserCertsRequest, clock clockwork.Clock) error {
+var errUserSingleUseUserCertNotNeeded = errors.New("single-use user certificate not needed to access this resource")
+
+// validateUserSingleUseCertRequest validates the request for a single-use user
+// cert. Returns nil if validation succeeds and MFA check is required, or
+// errUserSingleUseUserCertNotNeeded if validation succeeds and MFA
+// authentication is not required for this session. Any other error means
+// failed validation.
+func validateUserSingleUseCertRequest(ctx context.Context, actx *grpcContext, req *proto.UserCertsRequest) error {
+	if req.Username != actx.User.GetName() {
+		return trace.BadParameter("cannot request a single-use certificate for a different user")
+	}
+
+	notFoundErr := trace.NotFound("target resource not found")
+	var noMFAAccessErr error
 	switch req.Usage {
 	case proto.UserCertsRequest_SSH:
 		if req.NodeName == "" {
 			return trace.BadParameter("missing NodeName field in a ssh-only UserCertsRequest")
 		}
+		// Find the target node and check whether MFA is required.
+		nodes, err := actx.authServer.GetNodes(defaults.Namespace, services.SkipValidation())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		var node types.Server
+		for _, n := range nodes {
+			// Get the server address without port number.
+			addr, _, err := net.SplitHostPort(n.GetAddr())
+			if err != nil {
+				addr = n.GetAddr()
+			}
+			// Match NodeName to UUID, hostname or self-reported server address.
+			if n.GetName() == req.NodeName || n.GetHostname() == req.NodeName || addr == req.NodeName {
+				node = n
+				break
+			}
```

Comment:
`n.GetName()` returns the UUID, so the first condition in this `if` statement matches nodes by UUID

Reply:
Yeah, what I mean is that iteration shouldn't halt if a hostname/addr match is found because there might still be a node which matches UUID and that would need to take priority.

Reply:
If some node has a hostname matching other node's UUID (which seems very unlikely), then we have multiple matches and no way to resolve the conflict and for the user to pick one of them.

I think the more likely issue is when user passes a hostname and multiple nodes match.
Changed the logic to catch that scenario - if there are multiple matches return an error.